accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1212 @@
+argument list `args1`. Then it executes the body of _Dm_ in an
+environment where `this` and the name of the representation are bound
+to `o`, the type variables of `V` are bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>,
+and the formal parameters of `m` are bound to `args1` in the same way
+that they would be bound for a normal function call.
+If the body completes returning an object `o2` then the invocation
+evaluates to `o2`. If the body throws an object and a stack trace
+then the invocation completes throwing the same object and stack
+trace.
+
+
+## Static Analysis of Extension Types
+
+Assume that
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>
+are types, and `V` resolves to an extension type declaration of the
+following form:
+
+```dart
+extension type V<X1 extends B1, .. Xs extends Bs>(T id) ... {
+  ... // Members.
+}
+```
+
+It is then allowed to use
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>
+as a type.
+
+*For example, it can occur as the declared type of a variable or parameter,
+as the return type of a function or getter, as a type argument in a type,
+as the representation type of an extension type declaration, as the on-type
+of an extension declaration, as the type in the `onPart` of a try/catch
+statement, in a type test `o is V<...>`, in a type cast `o as V<...>`, or
+as the body of a type alias. It is also allowed to create a new instance
+where one or more extension types occur as type arguments (e.g.,
+`List<V>.empty()`).*
+
+A compile-time error occurs if the type
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>
+is not regular-bounded.
+
+*In other words, such types can not be super-bounded. The reason for this
+restriction is that it is unsound to execute code in the body of `V` in
+the case where the values of the type variables do not satisfy their
+declared bounds, and those values will be obtained directly from the static
+type of the receiver in each member invocation on `V`.*
+
+A compile-time error occurs if a type parameter of an extension type
+declaration occurs in a non-covariant position in the representation type.
+
+When `s` is zero,
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>
+simply stands for `V`, a non-generic extension type.
+When `s` is greater than zero, a raw occurrence `V` is treated like a raw
+type: Instantiation to bound is used to obtain the omitted type arguments.
+*Note that this may yield a super-bounded type, which is then a
+compile-time error.*
+
+We say that the static type of said variable, parameter, etc.
+_is the extension type_
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and that its static type _is an extension type_.
+
+It is a compile-time error if `await e` occurs, and the static type of
+`e` is an extension type which is not a subtype of `Future<T>` or
+`FutureOr<T>` for any `T`.
+
+A compile-time error occurs if an extension type declares a member whose
+name is declared by `Object` as well.
```

Comment:
-> is declared as an *instance member* by `Object` as well.

(`Object` has static `hash`-functions too.)

Do we allow same base-name, `set hashCode`? (Shouldn't be a problem, just asking.)

This precludes both static and instance members with those names, which is probably the right choice.

Also need to disallow the representation object field being named like that.
Maybe also constructors (but not actually sure, those were always quirky).

Reply:
Following extension methods, I made this an error when it is the same basename. Can relax later.

Added commentary about the representation name, repeating that it counts as an instance variable declaration during static analysis.